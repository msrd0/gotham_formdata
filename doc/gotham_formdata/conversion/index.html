<!DOCTYPE html><html lang="en"><head><meta charset="utf-8"><meta name="viewport" content="width=device-width, initial-scale=1.0"><meta name="generator" content="rustdoc"><meta name="description" content="This mod contains conversion traits for common used types, that allows them to be created from a stream of bytes. Furthermore, it allows every type that implements [FromStr] plus some other common types to be converted."><meta name="keywords" content="rust, rustlang, rust-lang, conversion"><title>gotham_formdata::conversion - Rust</title><link rel="stylesheet" type="text/css" href="../../normalize.css"><link rel="stylesheet" type="text/css" href="../../rustdoc.css" id="mainThemeStyle"><link rel="stylesheet" type="text/css" href="../../light.css"  id="themeStyle"><link rel="stylesheet" type="text/css" href="../../dark.css" disabled ><link rel="stylesheet" type="text/css" href="../../ayu.css" disabled ><script id="default-settings" ></script><script src="../../storage.js"></script><script src="../../crates.js"></script><noscript><link rel="stylesheet" href="../../noscript.css"></noscript><link rel="icon" type="image/svg+xml" href="../../favicon.svg"><link rel="alternate icon" type="image/png" href="../../favicon-16x16.png"><link rel="alternate icon" type="image/png" href="../../favicon-32x32.png"><style type="text/css">#crate-search{background-image:url("../../down-arrow.svg");}</style></head><body class="rustdoc mod"><!--[if lte IE 11]><div class="warning">This old browser is unsupported and will most likely display funky things.</div><![endif]--><nav class="sidebar"><div class="sidebar-menu" role="button">&#9776;</div><a href='../../gotham_formdata/index.html'><div class='logo-container rust-logo'><img src='../../rust-logo.png' alt='logo'></div></a><h2 class="location">Module conversion</h2><div class="sidebar-elems"><div class="block items"><ul><li><a href="#traits">Traits</a></li><li><a href="#types">Type Definitions</a></li></ul></div><div id="sidebar-vars" data-name="conversion" data-ty="mod" data-relpath="./"></div><script defer src="./sidebar-items.js"></script></div></nav><div class="theme-picker"><button id="theme-picker" aria-label="Pick another theme!" aria-haspopup="menu" title="themes"><img width="18" height="18" alt="Pick another theme!" src="../../brush.svg"></button><div id="theme-choices" role="menu"></div></div><nav class="sub"><form class="search-form"><div class="search-container"><div><select id="crate-search"><option value="All crates">All crates</option></select><input class="search-input"name="search" disabled autocomplete="off" spellcheck="false" placeholder="Click or press ‘S’ to search, ‘?’ for more options…" type="search"></div><button type="button" id="help-button" title="help">?</button><a id="settings-menu" href="../../settings.html" title="settings"><img width="18" height="18" alt="Change settings" src="../../wheel.svg"></a></div></form></nav><section id="main" class="content"><h1 class="fqn"><span class="in-band">Module <a href="../index.html">gotham_formdata</a>::<wbr><a class="mod" href="#">conversion</a><button id="copy-path" onclick="copy_path(this)" title="copy path"><img src="../../clipboard.svg" width="19" height="18" alt="Copy item import" title="Copy item import to clipboard"></button></span><span class="out-of-band"><span id="render-detail"><a id="toggle-all-docs" href="javascript:void(0)" title="collapse all docs">[<span class="inner">&#x2212;</span>]</a></span><a class="srclink" href="../../src/gotham_formdata/conversion.rs.html#1-173" title="goto source code">[src]</a></span></h1><details class="rustdoc-toggle top-doc" open><summary class="hideme"><span>Expand description</span></summary><div class="docblock"><p>This mod contains conversion traits for common used types, that allows them to be created from
a stream of bytes. Furthermore, it allows every type that implements <a href="https://doc.rust-lang.org/1.55.0/core/str/traits/trait.FromStr.html" title="FromStr">FromStr</a> plus some other
common types to be converted.</p>
<p>You cannot implement any of these traits manually. This is intentional. Instead, if you want to
provide a custom conversion method, just implement it as a method for your type:</p>

<div class="example-wrap"><pre class="rust rust-example-rendered">
<span class="kw">use</span> <span class="ident">futures_util</span>::{<span class="ident">FutureExt</span>, <span class="ident">StreamExt</span>};
<span class="kw">use</span> <span class="ident">gotham_formdata</span>::{<span class="ident">conversion::ConversionFuture</span>, <span class="ident">value</span>::{<span class="ident">BytesOrString</span>, <span class="ident">Value</span>}, <span class="ident">FormData</span>};

<span class="doccomment">/// This type parses Base64-encoded values to a [Vec&lt;u8&gt;].</span>
<span class="kw">struct</span> <span class="ident">Base64</span>(<span class="ident">Vec</span><span class="op">&lt;</span><span class="ident">u8</span><span class="op">&gt;</span>);

<span class="kw">impl</span> <span class="ident">Base64</span> {
	<span class="comment">// the method signature needs to be roughly equivalent to this</span>
	<span class="kw">async</span> <span class="kw">fn</span> <span class="ident">convert_value</span><span class="op">&lt;</span><span class="ident">E</span><span class="op">&gt;</span>(
			<span class="ident">name</span>: <span class="kw-2">&amp;</span><span class="ident">str</span>,
			<span class="ident">value</span>: <span class="ident">Value</span><span class="op">&lt;</span><span class="lifetime">&#39;_</span>, <span class="ident">gotham_formdata::Error</span><span class="op">&lt;</span><span class="ident">E</span><span class="op">&gt;</span><span class="op">&gt;</span>
	) <span class="op">-</span><span class="op">&gt;</span> <span class="prelude-ty">Result</span><span class="op">&lt;</span><span class="self">Self</span>, <span class="ident">gotham_formdata::Error</span><span class="op">&lt;</span><span class="ident">E</span><span class="op">&gt;</span><span class="op">&gt;</span>
	<span class="kw">where</span>
		<span class="ident">E</span>: <span class="ident">std::error::Error</span>
	{
		<span class="kw">let</span> <span class="ident">decoded</span> <span class="op">=</span> <span class="kw">match</span> <span class="ident">value</span>.<span class="ident">value</span> {
			<span class="ident">BytesOrString::Bytes</span>(<span class="kw-2">mut</span> <span class="ident">stream</span>) <span class="op">=</span><span class="op">&gt;</span> {
				<span class="kw">let</span> <span class="kw-2">mut</span> <span class="ident">buf</span>: <span class="ident">Vec</span><span class="op">&lt;</span><span class="ident">u8</span><span class="op">&gt;</span> <span class="op">=</span> <span class="ident">Vec::new</span>();
				<span class="kw">while</span> <span class="kw">let</span> <span class="prelude-val">Some</span>(<span class="ident">data</span>) <span class="op">=</span> <span class="ident">stream</span>.<span class="ident">next</span>().<span class="kw">await</span> {
					<span class="ident">buf</span>.<span class="ident">extend_from_slice</span>(<span class="kw-2">&amp;</span><span class="ident">data</span><span class="question-mark">?</span>);
				}
				<span class="ident">base64::decode</span>(<span class="kw-2">&amp;</span><span class="ident">buf</span>)
			},
			<span class="ident">BytesOrString::String</span>(<span class="ident">string</span>) <span class="op">=</span><span class="op">&gt;</span> <span class="ident">base64::decode</span>(<span class="ident">string</span>.<span class="ident">as_bytes</span>())
		}.<span class="ident">map_err</span>(<span class="op">|</span><span class="ident">err</span><span class="op">|</span> <span class="ident">gotham_formdata::Error::IllegalField</span>(<span class="ident">name</span>.<span class="ident">to_owned</span>(), <span class="ident">err</span>.<span class="ident">into</span>()))<span class="question-mark">?</span>;

		<span class="prelude-val">Ok</span>(<span class="self">Self</span>(<span class="ident">decoded</span>))
	}
}

<span class="attribute">#[<span class="ident">derive</span>(<span class="ident">FormData</span>)]</span>
<span class="kw">struct</span> <span class="ident">MyData</span> {
	<span class="ident">foo</span>: <span class="ident">Base64</span>
}</pre></div>
</div></details><h2 id="traits" class="section-header"><a href="#traits">Traits</a></h2>
<div class="item-table"><div class="item-left module-item"><a class="trait" href="trait.ConvertFromStr.html" title="gotham_formdata::conversion::ConvertFromStr trait">ConvertFromStr</a></div><div class="item-right docblock-short"><p>This trait is used to convert types that implement <a href="https://doc.rust-lang.org/1.55.0/core/str/traits/trait.FromStr.html" title="FromStr">FromStr</a> from a stream of bytes.</p>
</div><div class="item-left module-item"><a class="trait" href="trait.ConvertRawBytes.html" title="gotham_formdata::conversion::ConvertRawBytes trait">ConvertRawBytes</a></div><div class="item-right docblock-short"><p>This trait is used to convert <code>Vec&lt;u8&gt;</code> and similar types from a stream of bytes.</p>
</div></div><h2 id="types" class="section-header"><a href="#types">Type Definitions</a></h2>
<div class="item-table"><div class="item-left module-item"><a class="type" href="type.ConversionFuture.html" title="gotham_formdata::conversion::ConversionFuture type">ConversionFuture</a></div><div class="item-right docblock-short"><p>The future returned from conversion methods.</p>
</div></div></section><section id="search" class="content hidden"></section><div id="rustdoc-vars" data-root-path="../../" data-current-crate="gotham_formdata" data-search-index-js="../../search-index.js" data-search-js="../../search.js"></div>
    <script src="../../main.js"></script>
</body></html>