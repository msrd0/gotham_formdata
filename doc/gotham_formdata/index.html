<!DOCTYPE html><html lang="en"><head><meta charset="utf-8"><meta name="viewport" content="width=device-width, initial-scale=1.0"><meta name="generator" content="rustdoc"><meta name="description" content="This crate is an extension to the popular gotham web framework for Rust. It aims to reduce boilerplate necessary to read request bodies today as a stop-gap until gotham finally implements a body extractor."><meta name="keywords" content="rust, rustlang, rust-lang, gotham_formdata"><title>gotham_formdata - Rust</title><link rel="stylesheet" type="text/css" href="../normalize.css"><link rel="stylesheet" type="text/css" href="../rustdoc.css" id="mainThemeStyle"><link rel="stylesheet" type="text/css" href="../light.css"  id="themeStyle"><link rel="stylesheet" type="text/css" href="../dark.css" disabled ><link rel="stylesheet" type="text/css" href="../ayu.css" disabled ><script id="default-settings"></script><script src="../storage.js"></script><script src="../crates.js"></script><noscript><link rel="stylesheet" href="../noscript.css"></noscript><link rel="icon" type="image/svg+xml" href="../favicon.svg">
<link rel="alternate icon" type="image/png" href="../favicon-16x16.png">
<link rel="alternate icon" type="image/png" href="../favicon-32x32.png"><style type="text/css">#crate-search{background-image:url("../down-arrow.svg");}</style></head><body class="rustdoc mod"><!--[if lte IE 8]><div class="warning">This old browser is unsupported and will most likely display funky things.</div><![endif]--><nav class="sidebar"><div class="sidebar-menu" role="button">&#9776;</div><a href='../gotham_formdata/index.html'><div class='logo-container rust-logo'><img src='../rust-logo.png' alt='logo'></div></a><p class="location">Crate gotham_formdata</p><div class="block version"><p>Version 0.0.4</p></div><div class="sidebar-elems"><a id="all-types" href="all.html"><p>See all gotham_formdata's items</p></a><div class="block items"><ul><li><a href="#modules">Modules</a></li><li><a href="#enums">Enums</a></li><li><a href="#traits">Traits</a></li><li><a href="#types">Type Definitions</a></li></ul></div><p class="location"></p><div id="sidebar-vars" data-name="gotham_formdata" data-ty="mod" data-relpath="../"></div></div></nav><div class="theme-picker"><button id="theme-picker" aria-label="Pick another theme!" aria-haspopup="menu"><img src="../brush.svg" width="18" height="18" alt="Pick another theme!"></button><div id="theme-choices" role="menu"></div></div><script src="../theme.js"></script><nav class="sub"><form class="search-form"><div class="search-container"><div><select id="crate-search"><option value="All crates">All crates</option></select><input class="search-input" name="search" disabled autocomplete="off" spellcheck="false" placeholder="Click or press ‘S’ to search, ‘?’ for more options…" type="search"></div><button type="button" class="help-button">?</button>
                <a id="settings-menu" href="../settings.html"><img src="../wheel.svg" width="18" height="18" alt="Change settings"></a></div></form></nav><section id="main" class="content"><h1 class="fqn"><span class="in-band">Crate <a class="mod" href="">gotham_formdata</a></span><span class="out-of-band"><span id="render-detail"><a id="toggle-all-docs" href="javascript:void(0)" title="collapse all docs">[<span class="inner">&#x2212;</span>]</a></span><a class="srclink" href="../src/gotham_formdata/lib.rs.html#1-105" title="goto source code">[src]</a></span></h1><div class="docblock"><p>This crate is an extension to the popular <a href="https://github.com/gotham-rs/gotham">gotham web framework</a> for Rust. It aims to
reduce boilerplate necessary to read request bodies today as a stop-gap until gotham finally
implements a <a href="https://github.com/gotham-rs/gotham/issues/11">body extractor</a>.</p>
<h1 id="sparkles-features" class="section-header"><a href="#sparkles-features">:sparkles: Features</a></h1>
<ul>
<li>Parse <code>application/x-www-form-urlencoded</code> request bodies</li>
<li>Parse <code>multipart/form-data</code> request bodies</li>
<li>Verify the parsed request body</li>
<li><code>#![forbid(unsafe_code)]</code> ensures that all functionality is implemented in 100% safe Rust code</li>
</ul>
<h1 id="warning-warning" class="section-header"><a href="#warning-warning">:warning: Warning</a></h1>
<p>This crate is asynchronous, but does not yet enforce uploads limits. <strong>YOU ARE RESPONSIBLE
FOR ENFORCING UPLOAD LIMITS.</strong></p>
<h1 id="spiral_notepad-example" class="section-header"><a href="#spiral_notepad-example">:spiral_notepad: Example</a></h1>
<div class="example-wrap"><pre class="rust rust-example-rendered">
<span class="kw">use</span> <span class="ident">gotham_formdata</span>::<span class="ident">FormData</span>;

<span class="attribute">#[<span class="ident">derive</span>(<span class="ident">FormData</span>)]</span>
<span class="kw">struct</span> <span class="ident">LoginData</span> {
	<span class="attribute">#[<span class="ident">validate</span>(<span class="ident">regex</span> <span class="op">=</span> <span class="string">&quot;[a-zA-Z0-9_]&quot;</span>, <span class="ident">error</span> <span class="op">=</span> <span class="string">&quot;The username contains illegal characters.&quot;</span>)]</span>
	<span class="ident">username</span>: <span class="ident">String</span>,
	<span class="attribute">#[<span class="ident">validate</span>(<span class="ident">min_length</span> <span class="op">=</span> <span class="number">6</span>)]</span>
	<span class="ident">password</span>: <span class="ident">String</span>
}

<span class="kw">async</span> <span class="kw">fn</span> <span class="ident">login_handler</span>(<span class="ident">state</span>: <span class="kw-2">&amp;</span><span class="kw-2">mut</span> <span class="ident">State</span>) <span class="op">-</span><span class="op">&gt;</span> <span class="prelude-ty">Result</span><span class="op">&lt;</span><span class="ident">Response</span><span class="op">&lt;</span><span class="ident">Body</span><span class="op">&gt;</span>, <span class="ident">HandlerError</span><span class="op">&gt;</span> {
	<span class="kw">let</span> <span class="ident">login_data</span>: <span class="ident">LoginData</span> <span class="op">=</span> <span class="ident">FormData</span>::<span class="ident">parse_form_data</span>(<span class="ident">state</span>).<span class="kw">await</span><span class="question-mark">?</span>;
	<span class="prelude-val">Ok</span>(<span class="kw">if</span> <span class="ident">login_data</span>.<span class="ident">password</span> <span class="op">=</span><span class="op">=</span> <span class="string">&quot;secret&quot;</span> {
		<span class="ident">create_response</span>(<span class="ident">state</span>, <span class="ident">StatusCode</span>::<span class="ident">OK</span>, <span class="ident">TEXT_PLAIN</span>, <span class="ident">login_data</span>.<span class="ident">username</span>)
	} <span class="kw">else</span> {
		<span class="ident">create_empty_response</span>(<span class="ident">state</span>, <span class="ident">StatusCode</span>::<span class="ident">FORBIDDEN</span>)
	})
}</pre></div>
<h1 id="label-versioning" class="section-header"><a href="#label-versioning">:label: Versioning</a></h1>
<p>Like all rust crates, this crate will follow semantic versioning guidelines. However, changing
the MSRV (minimum supported rust version) is not considered a breaking change.</p>
<h1 id="page_with_curl-license" class="section-header"><a href="#page_with_curl-license">:page_with_curl: License</a></h1>
<pre><code class="language-text">Licensed under the Apache License, Version 2.0 (the &quot;License&quot;);
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

	https://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an &quot;AS IS&quot; BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
</code></pre>
</div><h2 id="modules" class="section-header"><a href="#modules">Modules</a></h2>
<table><tr class="module-item"><td><a class="mod" href="conversion/index.html" title="gotham_formdata::conversion mod">conversion</a></td><td class="docblock-short"><p>This mod contains conversion traits for common used types, that allows them to be created from
a stream of bytes. Furthermore, it allows every type that implements <a href="https://doc.rust-lang.org/nightly/core/str/traits/trait.FromStr.html" title="FromStr">FromStr</a> plus some other
common types to be converted.</p>
</td></tr><tr class="module-item"><td><a class="mod" href="validate/index.html" title="gotham_formdata::validate mod">validate</a></td><td class="docblock-short"><p>This mod contains the <a href="../gotham_formdata/validate/trait.Validator.html" title="Validator">Validator</a> trait as well as pre-defined validation methods.</p>
</td></tr><tr class="module-item"><td><a class="mod" href="value/index.html" title="gotham_formdata::value mod">value</a></td><td class="docblock-short"><p>This mod defines the <a href="../gotham_formdata/value/struct.Value.html" title="Value">Value</a> used for parsing the form data.</p>
</td></tr></table><h2 id="enums" class="section-header"><a href="#enums">Enums</a></h2>
<table><tr class="module-item"><td><a class="enum" href="enum.Error.html" title="gotham_formdata::Error enum">Error</a></td><td class="docblock-short"><p>This error type is used when parsing form data from a request body was unsuccessful.</p>
</td></tr></table><h2 id="traits" class="section-header"><a href="#traits">Traits</a></h2>
<table><tr class="module-item"><td><a class="trait" href="trait.FormData.html" title="gotham_formdata::FormData trait">FormData</a></td><td class="docblock-short"><p>This is the trait implemented by <code>#[derive(FormData)]</code>. It provides a method to parse the struct
it is implemented for to be parsed from the request body contained in gotham’s state.</p>
</td></tr><tr class="module-item"><td><a class="trait" href="trait.FormDataFromState.html" title="gotham_formdata::FormDataFromState trait">FormDataFromState</a></td><td class="docblock-short"><p>This is the equivalent of <a href="../gotham_formdata/trait.FormData.html" title="FormData">FormData</a> from the state’s perspective. Use this if you prefer
<code>state.parse_form_data::&lt;MyData&gt;()?</code> over <code>MyData::parse_form_data(&amp;mut state)?</code>.</p>
</td></tr></table><h2 id="types" class="section-header"><a href="#types">Type Definitions</a></h2>
<table><tr class="module-item"><td><a class="type" href="type.FormDataFuture.html" title="gotham_formdata::FormDataFuture type">FormDataFuture</a></td><td class="docblock-short"><p>This is the return type of <a href="../gotham_formdata/trait.FormData.html#tymethod.parse_form_data" title="FormData::parse_form_data">FormData::parse_form_data</a>.</p>
</td></tr></table><h2 id="derives" class="section-header"><a href="#derives">Derive Macros</a></h2>
<table><tr class="module-item"><td><a class="derive" href="derive.FormData.html" title="gotham_formdata::FormData derive">FormData</a></td><td class="docblock-short"><p>This derive macro implements <code>FormData</code> for the struct it is invoked on. Enums, unions and
tuple structs are not supported.</p>
</td></tr></table></section><section id="search" class="content hidden"></section><section class="footer"></section><div id="rustdoc-vars" data-root-path="../" data-current-crate="gotham_formdata" data-search-js="../search-index.js"></div>
    <script src="../main.js"></script></body></html>