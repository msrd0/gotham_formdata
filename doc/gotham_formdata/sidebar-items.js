initSidebarItems({"derive":[["FormData","This derive macro implements `FormData` for the struct it is invoked on. Enums, unions and tuple structs are not supported."]],"enum":[["Error","This error type is used when parsing form data from a request body was unsuccessful."]],"mod":[["conversion","This mod contains conversion traits for common used types, that allows them to be created from a stream of bytes. Furthermore, it allows every type that implements [FromStr] or [From<&[u8]>] to be converted."],["validate","This mod contains the [Validator] trait as well as pre-defined validation methods."]],"trait":[["FormData","This is the trait implemented by `#[derive(FormData)]`. It provides a method to parse the struct it is implemented for to be parsed from the request body contained in gotham's state."],["FormDataFromState","This is the equivalent of [FormData] from the state's perspective. Use this if you prefer `state.parse_form_data::<MyData>()?` over `MyData::parse_form_data(&mut state)?`."]],"type":[["FormDataFuture","This is the return type of [FormData::parse_form_data]."]]});