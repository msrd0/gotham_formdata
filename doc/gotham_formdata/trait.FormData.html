<!DOCTYPE html><html lang="en"><head><meta charset="utf-8"><meta name="viewport" content="width=device-width, initial-scale=1.0"><meta name="generator" content="rustdoc"><meta name="description" content="This is the trait implemented by `#[derive(FormData)]`. It provides a method to parse the struct it is implemented for to be parsed from the request body contained in gotham’s state."><meta name="keywords" content="rust, rustlang, rust-lang, FormData"><title>FormData in gotham_formdata - Rust</title><link rel="preload" as="font" type="font/woff2" crossorigin href="../SourceSerif4-Regular.ttf.woff2"><link rel="preload" as="font" type="font/woff2" crossorigin href="../FiraSans-Regular.woff2"><link rel="preload" as="font" type="font/woff2" crossorigin href="../FiraSans-Medium.woff2"><link rel="preload" as="font" type="font/woff2" crossorigin href="../SourceCodePro-Regular.ttf.woff2"><link rel="preload" as="font" type="font/woff2" crossorigin href="../SourceSerif4-Bold.ttf.woff2"><link rel="preload" as="font" type="font/woff2" crossorigin href="../SourceCodePro-Semibold.ttf.woff2"><link rel="stylesheet" href="../normalize.css"><link rel="stylesheet" href="../rustdoc.css" id="mainThemeStyle"><link rel="stylesheet" href="../ayu.css" disabled><link rel="stylesheet" href="../dark.css" disabled><link rel="stylesheet" href="../light.css" id="themeStyle"><script id="default-settings" ></script><script src="../storage.js"></script><script defer src="sidebar-items.js"></script><script defer src="../main.js"></script><noscript><link rel="stylesheet" href="../noscript.css"></noscript><link rel="alternate icon" type="image/png" href="../favicon-16x16.png"><link rel="alternate icon" type="image/png" href="../favicon-32x32.png"><link rel="icon" type="image/svg+xml" href="../favicon.svg"></head><body class="rustdoc trait"><!--[if lte IE 11]><div class="warning">This old browser is unsupported and will most likely display funky things.</div><![endif]--><nav class="mobile-topbar"><button class="sidebar-menu-toggle">&#9776;</button><a class="sidebar-logo" href="../gotham_formdata/index.html"><div class="logo-container"><img class="rust-logo" src="../rust-logo.svg" alt="logo"></div></a><h2></h2></nav><nav class="sidebar"><a class="sidebar-logo" href="../gotham_formdata/index.html"><div class="logo-container"><img class="rust-logo" src="../rust-logo.svg" alt="logo"></div></a><h2 class="location"><a href="#">FormData</a></h2><div class="sidebar-elems"><section><h3><a href="#required-associated-types">Required Associated Types</a></h3><ul class="block"><li><a href="#associatedtype.Err">Err</a></li></ul><h3><a href="#required-methods">Required Methods</a></h3><ul class="block"><li><a href="#tymethod.parse_form_data">parse_form_data</a></li></ul><h3><a href="#implementors">Implementors</a></h3></section><h2><a href="index.html">In gotham_formdata</a></h2></div></nav><main><div class="width-limiter"><nav class="sub"><form class="search-form"><span></span><input class="search-input" name="search" autocomplete="off" spellcheck="false" placeholder="Click or press ‘S’ to search, ‘?’ for more options…" type="search"><div id="help-button" title="help" tabindex="-1"><a href="../help.html">?</a></div><div id="settings-menu" tabindex="-1"><a href="../settings.html" title="settings"><img width="22" height="22" alt="Change settings" src="../wheel.svg"></a></div></form></nav><section id="main-content" class="content"><div class="main-heading"><h1 class="fqn">Trait <a href="index.html">gotham_formdata</a>::<wbr><a class="trait" href="#">FormData</a><button id="copy-path" onclick="copy_path(this)" title="Copy item path to clipboard"><img src="../clipboard.svg" width="19" height="18" alt="Copy item path"></button></h1><span class="out-of-band"><a class="srclink" href="../src/gotham_formdata/form_data.rs.html#15-21">source</a> · <a id="toggle-all-docs" href="javascript:void(0)" title="collapse all docs">[<span class="inner">&#x2212;</span>]</a></span></div><div class="item-decl"><pre class="rust trait"><code>pub trait FormData: <a class="trait" href="https://doc.rust-lang.org/1.66.0/core/marker/trait.Sized.html" title="trait core::marker::Sized">Sized</a> {
    type <a href="#associatedtype.Err" class="associatedtype">Err</a>;

    fn <a href="#tymethod.parse_form_data" class="fnname">parse_form_data</a>(state: &amp;mut <a class="struct" href="../gotham/state/struct.State.html" title="struct gotham::state::State">State</a>) -&gt; <a class="type" href="type.FormDataFuture.html" title="type gotham_formdata::FormDataFuture">FormDataFuture</a>&lt;Self&gt;;
}</code></pre></div><details class="rustdoc-toggle top-doc" open><summary class="hideme"><span>Expand description</span></summary><div class="docblock"><p>This is the trait implemented by <code>#[derive(FormData)]</code>. It provides a method to parse the struct
it is implemented for to be parsed from the request body contained in gotham’s state.</p>
<p>You usually don’t implement this trait directly, use the derive macro instead.</p>
</div></details><h2 id="required-associated-types" class="small-section-header">Required Associated Types<a href="#required-associated-types" class="anchor"></a></h2><div class="methods"><details class="rustdoc-toggle method-toggle" open><summary><section id="associatedtype.Err" class="method has-srclink"><a class="srclink rightside" href="../src/gotham_formdata/form_data.rs.html#17">source</a><h4 class="code-header">type <a href="#associatedtype.Err" class="associatedtype">Err</a></h4></section></summary><div class="docblock"><p>The error type returned when parsing the request body was unsuccessful.</p>
</div></details></div><h2 id="required-methods" class="small-section-header">Required Methods<a href="#required-methods" class="anchor"></a></h2><div class="methods"><details class="rustdoc-toggle method-toggle" open><summary><section id="tymethod.parse_form_data" class="method has-srclink"><a class="srclink rightside" href="../src/gotham_formdata/form_data.rs.html#20">source</a><h4 class="code-header">fn <a href="#tymethod.parse_form_data" class="fnname">parse_form_data</a>(state: &amp;mut <a class="struct" href="../gotham/state/struct.State.html" title="struct gotham::state::State">State</a>) -&gt; <a class="type" href="type.FormDataFuture.html" title="type gotham_formdata::FormDataFuture">FormDataFuture</a>&lt;Self&gt;</h4></section></summary><div class="docblock"><p>Parse the struct from the request body contained in gotham’s state.</p>
</div></details></div><h2 id="implementors" class="small-section-header">Implementors<a href="#implementors" class="anchor"></a></h2><div id="implementors-list"></div><script src="../implementors/gotham_formdata/form_data/trait.FormData.js" async></script></section></div></main><div id="rustdoc-vars" data-root-path="../" data-current-crate="gotham_formdata" data-themes="ayu,dark,light" data-resource-suffix="" data-rustdoc-version="1.66.0 (69f9c33d7 2022-12-12)" ></div></body></html>