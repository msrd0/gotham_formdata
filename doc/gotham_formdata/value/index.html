<!DOCTYPE html><html lang="en"><head><meta charset="utf-8"><meta name="viewport" content="width=device-width, initial-scale=1.0"><meta name="generator" content="rustdoc"><meta name="description" content="This mod defines the Value used for parsing the form data."><title>gotham_formdata::value - Rust</title><link rel="preload" as="font" type="font/woff2" crossorigin href="../../static.files/SourceSerif4-Regular-46f98efaafac5295.ttf.woff2"><link rel="preload" as="font" type="font/woff2" crossorigin href="../../static.files/FiraSans-Regular-018c141bf0843ffd.woff2"><link rel="preload" as="font" type="font/woff2" crossorigin href="../../static.files/FiraSans-Medium-8f9a781e4970d388.woff2"><link rel="preload" as="font" type="font/woff2" crossorigin href="../../static.files/SourceCodePro-Regular-562dcc5011b6de7d.ttf.woff2"><link rel="preload" as="font" type="font/woff2" crossorigin href="../../static.files/SourceCodePro-Semibold-d899c5a5c4aeb14a.ttf.woff2"><link rel="stylesheet" href="../../static.files/normalize-76eba96aa4d2e634.css"><link rel="stylesheet" href="../../static.files/rustdoc-9ee3a5e31a2afa3e.css"><meta name="rustdoc-vars" data-root-path="../../" data-static-root-path="../../static.files/" data-current-crate="gotham_formdata" data-themes="" data-resource-suffix="" data-rustdoc-version="1.75.0 (82e1608df 2023-12-21)" data-channel="1.75.0" data-search-js="search-8fbf244ebcf71464.js" data-settings-js="settings-74424d7eec62a23e.js" ><script src="../../static.files/storage-fec3eaa3851e447d.js"></script><script defer src="../sidebar-items.js"></script><script defer src="../../static.files/main-9dd44ab47b99a0fb.js"></script><noscript><link rel="stylesheet" href="../../static.files/noscript-5d8b3c7633ad77ba.css"></noscript><link rel="alternate icon" type="image/png" href="../../static.files/favicon-16x16-8b506e7a72182f1c.png"><link rel="alternate icon" type="image/png" href="../../static.files/favicon-32x32-422f7d1d52889060.png"><link rel="icon" type="image/svg+xml" href="../../static.files/favicon-2c020d218678b618.svg"></head><body class="rustdoc mod"><!--[if lte IE 11]><div class="warning">This old browser is unsupported and will most likely display funky things.</div><![endif]--><nav class="mobile-topbar"><button class="sidebar-menu-toggle">&#9776;</button></nav><nav class="sidebar"><div class="sidebar-crate"><h2><a href="../../gotham_formdata/index.html">gotham_formdata</a><span class="version">0.0.7</span></h2></div><h2 class="location"><a href="#">Module value</a></h2><div class="sidebar-elems"><section><ul class="block"><li><a href="#structs">Structs</a></li><li><a href="#enums">Enums</a></li><li><a href="#types">Type Aliases</a></li></ul></section><h2><a href="../index.html">In crate gotham_formdata</a></h2></div></nav><main><div class="width-limiter"><nav class="sub"><form class="search-form"><span></span><input class="search-input" name="search" aria-label="Run search in the documentation" autocomplete="off" spellcheck="false" placeholder="Click or press ‘S’ to search, ‘?’ for more options…" type="search"><div id="help-button" title="help" tabindex="-1"><a href="../../help.html">?</a></div><div id="settings-menu" tabindex="-1"><a href="../../settings.html" title="settings"><img width="22" height="22" alt="Change settings" src="../../static.files/wheel-7b819b6101059cd0.svg"></a></div></form></nav><section id="main-content" class="content"><div class="main-heading"><h1>Module <a href="../index.html">gotham_formdata</a>::<wbr><a class="mod" href="#">value</a><button id="copy-path" title="Copy item path to clipboard"><img src="../../static.files/clipboard-7571035ce49a181d.svg" width="19" height="18" alt="Copy item path"></button></h1><span class="out-of-band"><a class="src" href="../../src/gotham_formdata/value.rs.html#1-42">source</a> · <button id="toggle-all-docs" title="collapse all docs">[<span>&#x2212;</span>]</button></span></div><details class="toggle top-doc" open><summary class="hideme"><span>Expand description</span></summary><div class="docblock"><p>This mod defines the <a href="struct.Value.html" title="struct gotham_formdata::value::Value">Value</a> used for parsing the form data.</p>
</div></details><h2 id="structs" class="small-section-header"><a href="#structs">Structs</a></h2><ul class="item-table"><li><div class="item-name"><a class="struct" href="struct.Mime.html" title="struct gotham_formdata::value::Mime">Mime</a></div><div class="desc docblock-short">A parsed mime or media type.</div></li><li><div class="item-name"><a class="struct" href="struct.Value.html" title="struct gotham_formdata::value::Value">Value</a></div><div class="desc docblock-short">A value, either a string parsed from and urlencoded form, or bytes from a multipart body.</div></li></ul><h2 id="enums" class="small-section-header"><a href="#enums">Enums</a></h2><ul class="item-table"><li><div class="item-name"><a class="enum" href="enum.BytesOrString.html" title="enum gotham_formdata::value::BytesOrString">BytesOrString</a></div><div class="desc docblock-short">Either a stream of bytes or a string.</div></li></ul><h2 id="types" class="small-section-header"><a href="#types">Type Aliases</a></h2><ul class="item-table"><li><div class="item-name"><a class="type" href="type.ByteStream.html" title="type gotham_formdata::value::ByteStream">ByteStream</a></div><div class="desc docblock-short">A stream of bytes.</div></li></ul></section></div></main></body></html>