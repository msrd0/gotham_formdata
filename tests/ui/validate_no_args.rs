use gotham_formdata::FormData;

#[derive(FormData)]
struct MyFormData {
	#[validate]
	value: u8
}

fn main() {
}
