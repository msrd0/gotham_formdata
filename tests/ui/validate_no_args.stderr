error: Illegal syntax for attribute validate
 --> $DIR/validate_no_args.rs:5:4
  |
5 |     #[validate]
  |       ^^^^^^^^
